### PR TITLE
Fix empty request bodies

### LINKED ISSUE #51: Fix empty method argument type definitions
i.e.

```
$ npm run dtslint types/gapi.client.adexchangebuyer

> google-api-typings-generator@ dtslint /Users/namoscato/dev/git/google-api-typings-generator
> dtslint "types/gapi.client.adexchangebuyer"

Error: /Users/namoscato/dev/git/google-api-typings-generator/types/gapi.client.adexchangebuyer/gapi.client.adexchangebuyer-tests.ts:35:52
ERROR: 35:52    expect  TypeScript@next compile error: 
Expected 1 arguments, but got 0.
ERROR: 83:55    expect  TypeScript@next compile error: 
Expected 1 arguments, but got 0.
ERROR: 128:60   expect  TypeScript@next compile error: 
Argument expression expected.
ERROR: 129:13   expect  TypeScript@next compile error: 
Argument of type '{ accountId: number; adChoicesDestinationUrl: string; adTechnologyProviders: { detectedProviderIds: string[]; hasUnidentifiedProvider: boolean; }; advertiserId: string[]; advertiserName: string; ... 25 more ...; width: number; }' is not assignable to parameter of type '{ alt?: string | undefined; fields?: string | undefined; key?: string | undefined; oauth_token?: string | undefined; prettyPrint?: boolean | undefined; quotaUser?: string | undefined; userIp?: string | undefined; resource: Creative; }'.
  Object literal may only specify known properties, and 'accountId' does not exist in type '{ alt?: string | undefined; fields?: string | undefined; key?: string | undefined; oauth_token?: string | undefined; prettyPrint?: boolean | undefined; quotaUser?: string | undefined; userIp?: string | undefined; resource: Creative; }'.
ERROR: 1082:60  expect  TypeScript@next compile error: 
Argument expression expected.
ERROR: 1083:13  expect  TypeScript@next compile error: 
Argument of type '{ proposals: { billedBuyer: { accountId: string; }; buyer: { accountId: string; }; buyerContacts: { email: string; name: string; }[]; buyerPrivateData: { referenceId: string; referencePayload: string; }; ... 18 more ...; sellerContacts: { ...; }[]; }[]; webPropertyCode: string; }' is not assignable to parameter of type '{ alt?: string | undefined; fields?: string | undefined; key?: string | undefined; oauth_token?: string | undefined; prettyPrint?: boolean | undefined; quotaUser?: string | undefined; userIp?: string | undefined; resource: CreateOrdersRequest; }'.
  Object literal may only specify known properties, and 'proposals' does not exist in type '{ alt?: string | undefined; fields?: string | undefined; key?: string | undefined; oauth_token?: string | undefined; prettyPrint?: boolean | undefined; quotaUser?: string | undefined; userIp?: string | undefined; resource: CreateOrdersRequest; }'.

    at /Users/namoscato/dev/git/google-api-typings-generator/node_modules/dtslint/bin/index.js:193:19
```

_Originally posted by @namoscato in https://github.com/Maxim-Mazurok/google-api-typings-generator/pull/42#issuecomment-583885619_

### PR DESCRIPTION
**Pull request by namoscato**
Fixes #51

* Make request argument optional for empty bodies
* Minimize duplication via RequestParameterWriterCallback
* Unit test getResourceTypeName
* Reverts https://github.com/Maxim-Mazurok/google-api-typings-generator/pull/52 `rootDir` change

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -491,73 +515,49 @@ export class App {
...
-          if (
-            !(requestParameters.hasOwnProperty('resource') && hasRequestRef)
-          ) {
-            // no resource param and no body at the same time -> generate x(request)
+
+          const requestRef = method.request?.$ref;
+          const requestParameters: Record<string, JsonSchema> = sortObject({
+            ...parameters,
+            ...method.parameters,
+          });
```
Comment (namoscato) [inline src/app.ts]: The current state assumes `requestParameters` is not empty, which might not be the case?

File: src/app.ts [hunk 2]
```diff
@@ -491,73 +515,49 @@ export class App {
...
-                        );
-                      }
-                    });
-                  },
+                  type: App.createRequestParameterWriterCallback(
+                    requestParameters,
+                    schemas,
+                    requestRef
+                  ),
+                  required: Boolean(requestRef),
```
Comment (namoscato) [inline src/app.ts]: This was one of the primary changes: The `request` argument should be optional if there is no (required) request body.

File: src/app.ts [hunk 3]
```diff
@@ -930,25 +928,31 @@ export class App {
...
         scope.scope(() => {
           this.writeProperties(scope, api, params);
         });
       }
-      const ref = method.request?.$ref;
-      if (ref != null) {
+
+      if (ref) {
+        if (!params) {
+          scope.write(`{} `);
```
Comment (namoscato) [inline src/app.ts]: Arguably not the best solution but ensures the first argument is always passed in, even if there are no method-level parameters.

File: src/app.ts (hunk 1 above)
Comment (namoscato) [reply src/app.ts]: This _does_ seem to be not empty across the board, so I think this is fine for now.

**Comment by Maxim-Mazurok** [comment]
Great job, thank you a lot!

**Pull request merged**